hw/ip/otbn/dv/otbnsim/sim/wsr.py

```diff
@@ -86,33 +86,54 @@ def changes(self) -> List[TraceWSR]:
 
 
 class RandWSR(WSR):
-    '''The magic RND WSR'''
+    '''The magic RND WSR
+
+    RND is special as OTBN can stall on reads to it. A read from RND either
+    immediately returns data from a cache of a previous EDN request (triggered
+    by writing to the RND_PREFETCH CSR) or waits for data from the EDN. To model
+    this anything reading from RND must first call `request_value` which returns
+    True if the value is available.
+    '''
     def __init__(self, name: str):
         super().__init__(name)
 
-        # For now, the RTL doesn't have a real "random number generator".
-        # Eventually, it will have an LFSR of some sort, seeded by the
-        # CSRNG/EDN. We'll model that properly when we've specced it out. Until
-        # then, random numbers are constant.  This constant must match the one
-        # in the RTL (the `rnd` signal in the `otbn_core` module found in
-        # rtl/otbn_core.sv). If changed here it must be changed there to match.
-        # Constant for RND is the binary bit pattern 1001 (0x9 hex) repeated to
-        # fill a 256-bit word.
-        u32 = 0x99999999
-        u64 = (u32 << 32) | u32
-        u128 = (u64 << 64) | u64
-        self._random_value = (u128 << 128) | u128
+        self._random_value = None # type: Optional[int]
+        self._random_value_read = False
+        self.pending_request = False
 
     def read_unsigned(self) -> int:
+        assert self._random_value is not None
+
+        self._random_value_read = True
+
         return self._random_value
 
     def read_u32(self) -> int:
         '''Read a 32-bit unsigned result'''
-        return self._random_value & ((1 << 32) - 1)
+        return self.read_unsigned() & ((1 << 32) - 1)
 
     def write_unsigned(self, value: int) -> None:
         return
 
+    def commit(self) -> None:
+        if self._random_value_read:
+            self._random_value = None
+            self.pending_request = False
```

Comment:
Is there ever a situation where `self.pending_request and self._random_value_read`? If I've understood the design properly, I don't think there is, in which case I think this should be an assertion instead of an assignment.

Reply:
Yes, that is the standard behaviour.

1. CSR/WSR instruction calls `request_value` in `pre_execute`, if no random value is available yet `pending_request` is set, instruction stalls and doesn't execute this cycle
2. Repeated `request_value` calls are made each step until the random value becomes available
3. Instruction actually executes and calls `read_unsigned`, `_random_value_read` is set to True
4. Following successful instruction execution commit happens, `_random_value_read` is set to True so we set `_random_value` to None (which will be an integer beforehand in this particular example) and `pending_request` is set to False (having been True).

`pending_request` is only set to False in 2 places, `__init__` and `commit` so if it's not getting set to False in `commit` we'd need to do it elsewhere and doing it on commit having seen a read seemed like the most sensible place to do it.

Reply:
Oh, I see! I didn't realise that we don't clear `pending_request` on a stalled instruction. Ok, sounds sensible.